idx/memory/tag_query.go

```diff
@@ -171,247 +239,812 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+		// whether those tags should be part of the final result set
+		if len(metricTags) > 0 {
+			if q.testByAllExpressions(id, def, matchName) {
+				for key := range metricTags {
+					select {
+					case tagCh <- key:
+					case <-stopCh:
+						// if execution of query has stopped because the max tag
+						// count has been reached then tagCh <- might block
+						// because that channel will not be consumed anymore. in
+						// that case the stop channel will have been closed so
+						// we so we exit here
+						break IDS
+					}
+					resultsCache[key] = struct{}{}
+				}
+			} else {
+				// check if we need to stop
+				select {
+				case <-stopCh:
+					break IDS
+				default:
+				}
+			}
+		}
+	}
+
+	q.wg.Done()
+	completeCh <- struct{}{}
+}
+
+// determines whether the given tag prefix/tag match will match the special
+// tag "name". if it does, then we can omit some filtering because we know
+// that every metric has a name
+func (q *TagQuery) tagFilterMatchesName() bool {
+	matchName := false
+
+	if q.filterTagBy == PREFIX_TAG || q.startWith == PREFIX_TAG {
+		if len(q.tagPrefix) <= 4 && "name"[:len(q.tagPrefix)] == q.tagPrefix {
+			matchName = true
+		}
+	} else if q.filterTagBy == MATCH_TAG || q.startWith == MATCH_TAG {
+		if q.tagMatch.value.MatchString("name") {
+			matchName = true
+		}
+	} else {
+		// some tag queries might have no prefix specified yet, in this case
+		// we do not need to filter by the name
+		matchName = true
+	}
+
+	return matchName
+}
+
+// RunGetTags executes the tag query and returns all the tags of the
+// resulting metrics
+func (q *TagQuery) RunGetTags(index TagIndex, byId map[string]*idx.Archive) map[string]struct{} {
+	q.index = index
+	q.byId = byId
+
+	maxTagCount := int32(math.MaxInt32)
+
+	// start a thread to calculate the maximum possible number of tags.
+	// this might not always complete before the query execution, but in most
+	// cases it likely will. when it does end before the execution of the query,
+	// the value of maxTagCount will be used to abort the query execution once
+	// the max number of possible tags has been reached
+	q.wg.Add(1)
+	go atomic.StoreInt32(&maxTagCount, int32(q.getMaxTagCount()))
+
+	q.sortByCost()
+	idCh, stopCh := q.getInitialIds()
+	completeCh := make(chan struct{}, tagQueryWorkers)
+	tagCh := make(chan string)
+
+	// matchName defines whether the given tag condition matches the special
+	// tag "name" or not
+	// if we know it matches "name", and we assume every metric has a name, then
+	// we can omit a lot of matching
+	matchName := q.tagFilterMatchesName()
+
+	// start the tag query workers. they'll consume the ids on the idCh and
+	// evaluate for each of them whether it satisfies all the conditions
+	// defined in the query expressions. then they will extract the tags of
+	// those that satisfy all conditions and push them into tagCh.
+	// when a worker completes it pushes an empty item into the completeCh to
+	// signal its completion
```

Comment:
there's no need for the completeCh stuff. a more common pattern to do this is:
```
go func() {
q.wg.Wait()
close(tagCh)
}()
```
then just do a for range over tagCh to consume everything.

Reply:
wouldn't multiple `close()` (one for each worker) result in panic? 
https://play.golang.org/p/mOVTPh7KYR

Reply:
If i get rid of the `completeCh`, how could the main thread then know that all workers have completed while still consuming the `tagCh` until it is empty? The workers can't close the `tagCh` when they are complete, because only the last one should close it.

Reply:
makes sense, will implement that

Reply:
that is much better: https://github.com/grafana/metrictank/pull/779/commits/8661c47647b98c4ee9fee08d30e57f603208c9fb

Reply:
yes, beautiful :)